Adds support for type parameter defaults

rbuckton opened the pull request:
This PR adds support for defaults for generic type parameters and replaces #6354 as it is fairly out of date. 

Default types for generic type parameters have the following syntax:

```grammarkdown
TypeParameter :
  BindingIdentifier Constraint? DefaultType?

DefaultType :
  `=` Type
```

Default types have the following rules:

* Required type parameters must not follow optional type parameters.
* When the checker applies defaults for type parameters, they are applied from left to right:
  * If the default for a type parameter refers to a type parameter that occurs earlier in the list, the type argument for that parameter will be used.
  * If the default for a type parameter refers to a type parameter that occurs later in the list, the empty object type `{}` will be used. 
  * *NOTE: This helps to avoid the cost of a circularity check for defaults.*
* Default types for a type parameter must satisfy the constraint for the type parameter, if it exists.
* A class or interface declaration that merges with an existing class or interface declaration may introduce a default for an existing type parameter.
* A class or interface declaration that merges with an existing class or interface declaration may introduce a new type parameter as long as it specifies a default.
* When resolving a symbol with the same name as a Type Parameter introduced in a merged declaration, the type parameter will not be considered and resolution will continue on to the next container.
* When specifying type arguments, you are only required to specify type arguments for the required type parameters. Unspecified type parameters will resolve to their default types.
* If a default type is not specified and inference cannot choose a candidate, the empty object type `{}` is inferred. *This is the existing behvaior.*
* If a default type is specified and inference cannot chose a candidate, the default type is inferred.

This PR also updates the signature for `Promise<T>` and `PromiseLike<T>` to take full advantage of the new support for type parameter defaults.

Fixes #2175
Fixes #10977
Fixes #12725 
Fixes #12886
Fixes #12910
Fixes #13008
Fixes #13015 
Fixes #13117

rbuckton commented:
@RyanCavanaugh: this is my take on refreshing the genericDefaults PR. It should resolve the open items you had on #6354 regarding checking defaults against constraints. I also added missing support for typeToString and the declaration emitter.  Please let me know if you can suggest any additional tests.

@mhegazy: when we discussed this, I suggested that interface declarations that have defaults must have identical defaults, allowing you to mix declarations with and without defaults:

```ts
interface X<A, B> {}
interface X<A, B = number> {}
```

This PR does not reflect that suggestion, and is instead more strict, requiring all declarations to have identical defaults as well. It seems more consistent, and we can loosen this restriction at any point if we so choose.

@ahejlsberg: The `getResolvedDefault` function in checker.ts is somewhat similar to `getResolvedBaseConstraint`. However, I chose to reuse the existing `pushTypeResolution`/`popTypeResolution` functions to perform circularity checks. Is there a specific advantage for `getResolvedBaseConstraint` to have its own circularity logic? Would you recommend I use that approach for `getResolvedDefault`, or would it make sense to modify `getResolvedBaseConstraint` to use the type-resolution functions?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -4824,6 +4840,45 @@ namespace ts {
+                const types = (<UnionOrIntersectionType>type).types;
+                const defaultTypes = filter(map(types, getResolvedDefaultWorker), x => x !== circularConstraintOrDefaultType);
+                return type.flags & TypeFlags.Union && defaultTypes.length === types.length ? getUnionType(defaultTypes) :
+                    type.flags & TypeFlags.Intersection && defaultTypes.length ? getIntersectionType(defaultTypes) :
```
This is a little confusing - so if there were any circular types in a union, you return `undefined`. But you disregard all circularities in an intersection type?

rbuckton replied:
I reused this from [line 4827](https://github.com/Microsoft/TypeScript/pull/13487/files/442f5408b6ecb4e4b34986bc161ed9037e5e61b5#diff-c3ed224e4daa84352f7f1abcd23e8ccaR4827), above. It does seem like it should also be performing a comparison against types.length as well, however I am unsure as to whether there was a specific reason for 4827 to ignore circularity in an intersection type.

rbuckton replied:
I made this more consistent in 15cb02e and I've added some additional circularity tests in ca16ba8.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -5106,6 +5161,53 @@ namespace ts {
+                        typeArguments = [];
+                    }
+                    const mapper: TypeMapper = t => {
+                        for (let i = 0; i < numTypeParameters; i++) {
```
You technically don't need to run through the initial type arguments that were already supplied, do you? Could you just start at `numTypeArguments` as you did below?

rbuckton replied:
No, if a type parameter has a default type that is a reference to another type parameter, then we need to map it to the other, possibly earlier, type parameter.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -5106,6 +5161,53 @@ namespace ts {
+                    }
+                    const mapper: TypeMapper = t => {
+                        for (let i = 0; i < numTypeParameters; i++) {
+                            if (t === typeParameters[i]) {
```
```ts
if (t !== typeParameters[i]) {
    continue;
}
```

rbuckton replied:
I have restructured and simplified this in 25cb02e

rbuckton commented:
> @mhegazy: when we discussed this, I suggested that interface declarations that have defaults must have identical defaults, allowing you to mix declarations with and without defaults...
> ...this PR does not reflect that suggestion and is more strict, requiring all declarations to have identical defaults as well. It seems more consistent and we can loosen this restriction at any point if we so choose.

The latest commit makes the following changes (also noted above):

> * <del>If any declaration of an interface or class has a default type for a type parameter, all declarations must have an identical default type.</del>
> * <ins>A class or interface declaration that merges with an existing class or interface declaration may introduce a default for an existing type parameter.</ins>
> * <ins>A class or interface declaration that merges with an existing class or interface declaration may introduce a new type parameter as long as it specifies a default.</ins>

For example:
```ts
interface X { } // no type parameters
interface X<T = never> { } // ok, add a new default type parameter

interface Y<T> { } // one required type parameter
interface Y<T = never> { } // ok, make an existing type parameter optional

interface Z { } // no type parameters
interface Z<T> { } // error, type parameter lists aren't identical

//// uncomment the following to remove the error as the declarations will merge:
// interface Z<T = never> { }
```

rbuckton commented:
@ahejlsberg Per our conversation in the design meeting, I've further modified `chooseOverload` to avoid additional inference overhead when all type arguments are supplied.

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -8983,8 +9155,15 @@ namespace ts {
+
         function getInferredType(context: InferenceContext, index: number): Type {
-            let inferredType = context.inferredTypes[index];
+            let inferredType = context.inferredTypes[index] || getSuppliedType(context, index);
```
Weird newline thing here

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -18205,16 +18409,110 @@ namespace ts {
+                    minTypeArgumentCount = maxTypeArgumentCount;
+                }
+                if (typeParameters && typeParameters.length > maxTypeArgumentCount) {
+                    // Trim the type parameters to the maximum length
```
What's an example of how this could happen?

rbuckton replied:
This is to preserve the current behavior with respect to where we report errors for type parameter lists:
```ts
interface X<T> {} // no error on first decl
interface X<T, U> {} // no defaults so we trim to `<T>` and report error for this decl
```

RyanCavanaugh replied:
Makes sense, thanks

ahejlsberg replied:
I don't think it works to trim like this. You might have members that reference the trimmed type parameter and those members would never be properly instantiated. In other words, the trimmed type parameters could "leak".

rbuckton replied:
The intent is that excess type parameters would result in an error, consistent with our behavior today as in the example above. Today we only get the type parameters of the first declaration, so this shouldn't be any different than how we would instantiate `X` above today.

RyanCavanaugh commented on src/compiler/diagnosticMessages.json:
```diff
@@ -2035,6 +2035,18 @@
         "category": "Error",
         "code": 2704
     },
+    "Required type parameters may not follow optional type parameters": {
```
Missing trailing period in this and in line 2046

rbuckton commented:
Following the discussion during the design meeting, I am considering making the following changes:

> * When specifying type arguments, you are only required to specify type arguments for the required type parameters. <del>Unspecified type parameters can be inferred.</del> <ins>Unspecified type parameters will resolve to their default types.</ins>

For example:

```ts
declare function f<T, U = T>(a: T, b?: U): void;

f(1); // ok. Using inference: T is number, U is number.
f(1, "a"); // ok. Using inference: T is number, U is string.
f<number>(1, "a"); // error. Not using inference: T is number, U is number. 
```

rbuckton commented:
I have added the following restriction:
> * <ins>It is an error to reference a type parameter in a class or interface declaration that is not specified in the declaration, but was introduced in a merged declaration.</ins>

This is to provide an error for the following case:

```ts
interface A { }
interface B {
  x: A; // error
}
interface B<A = number> {
}
```

DanielRosenwasser commented on src/compiler/parser.ts:
```diff
@@ -2102,6 +2103,10 @@ namespace ts {
 
+            if (parseOptional(SyntaxKind.EqualsToken)) {
+                node.default = parseType();
+            }
```
Nit: I would really love for if the parsing was order-agnostic for the purpose of error reporting, and then gave a grammar error later on. This could be done as a separate pass.

rbuckton commented:
@ahejlsberg I have updated the PR based on your offline feedback. I have removed the circularity check and have introduced mappers that only follow backward references for type parameters:

```ts
// backward reference
declare function f1<T = number, U = T>(): [T, U];
f1(); // infers '[number, number]'

// forward reference
declare function f2<T = U, U = number>(); [T, U];
f2(); // infers '[{}, number]'
```

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -4022,6 +4029,7 @@ namespace ts {
                     type.typeParameters = concatenate(outerTypeParameters, localTypeParameters);
                     type.outerTypeParameters = outerTypeParameters;
                     type.localTypeParameters = localTypeParameters;
+                    type.minTypeArgumentCount = getMinTypeArgumentCount(localTypeParameters);
```
Is there a reason we need the `minTypeArgumentCount` property. Could we just compute it on demand?

rbuckton replied:
Considering the number of times we need to check the arity of type arguments, it seemed worth the effort to pre-calculate this value just as we do `minArgumentCount` on signatures. Even if we compute on demand, I would likely want to cache the result to avoid repeatedly computing this value.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -5126,6 +5161,62 @@ namespace ts {
+         * @param minTypeArgumentCount The minimum number of required type arguments.
+         */
+        function fillMissingTypeArguments(typeArguments: Type[] | undefined, typeParameters: TypeParameter[] | undefined, minTypeArgumentCount: number) {
+            const numTypeParameters = typeParameters ? typeParameters.length : 0;
```
Use `length` function (in core.ts).

rbuckton replied:
This should be addressed throughout checker.ts as of 7616e37.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -5126,6 +5161,62 @@ namespace ts {
+                    // Map an unsatisfied type parameter with a default type.
+                    // If a type parameter does not have a default type, or if the default type
+                    // is a forward reference, the empty object type is used.
+                    const mapper: TypeMapper = t => {
```
A `TypeMapper` is also supposed to initialize the `mappedTypes` property. I think a better approach here is to back fill an array with `emptyObjectType` for the missing arguments and the call the regular `createTypeMapper`.

rbuckton replied:
This has been addressed as of e001258.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -9047,11 +9155,21 @@ namespace ts {
-                    inferredType = emptyObjectType;
+                    const defaultType = getDefaultFromTypeParameter(context.signature.typeParameters[index]);
+                    if (defaultType) {
+                        const backreferenceMapper: TypeMapper = t => indexOf(context.signature.typeParameters, t) >= index ? emptyObjectType : t;
```
See my previous comment about custom type mappers. Also, why do we need this `backreferenceMapper`?

rbuckton replied:
This was meant to only allow a type parameter default to point to an earlier type parameter, but not a later one. I will revise this based on your other comments.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -16052,9 +16176,13 @@ namespace ts {
                         const symbol = getNodeLinks(node).resolvedSymbol;
                         const typeParameters = symbol.flags & SymbolFlags.TypeAlias ? getSymbolLinks(symbol).typeParameters : (<TypeReference>type).target.localTypeParameters;
-                        checkTypeArgumentConstraints(typeParameters, node.typeArguments);
+                        const minTypeArgumentCount = symbol.flags & SymbolFlags.TypeAlias ? getSymbolLinks(symbol).minTypeArgumentCount : (<TypeReference>type).target.minTypeArgumentCount;
```
What's going on here?

rbuckton replied:
This is getting the pre-computed count of non-optional type arguments. As with `typeParameters` above, this is stored on `SymbolLinks` for a type alias, or the generic target for a `TypeReference`.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -16052,9 +16176,13 @@ namespace ts {
+                        checkTypeArgumentConstraints(typeParameters, node.typeArguments, minTypeArgumentCount);
                     }
                 }
+                if (type.flags & TypeFlags.TypeParameter && !(<TypeParameter>type).isThisType && type.symbol && !isTypeParameterInScope(<TypeParameter>type, node)) {
```
Would it be better not to consider type parameters in scope unless they are declared in that particular container type?

rbuckton replied:
That is what this is trying to accomplish. `isTypeParameterInScope` walks up the spine from `node` to find a declaration that contains the type parameter. Are you stating that this condition is too permissive or that there is a better approach?

ahejlsberg replied:
Here you are catching the situation after the fact and giving an error (which, although unlikely, *may* cause code to break). An alternative would be to add logic in `resolveName` to bypass type parameters that aren't declared by the enclosing construct. I was wondering if you considered that?

rbuckton replied:
I agree that this should move to `resolveName`. However, I don't believe we should bypass a type parameter that isn't declared locally but instead should report an error as we are doing here.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -18285,16 +18439,110 @@ namespace ts {
+            resolveTypeParametersOfClassOrInterface(symbol);
+
+            const { typeParameters, minTypeArgumentCount } = getSymbolLinks(symbol);
+            const maxTypeArgumentCount = typeParameters ? typeParameters.length : 0;
```
Use `length` function.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -4845,6 +4861,92 @@ namespace ts {
+                const defaultType = getDefaultFromTypeParameter(typeParameter);
+                const type = defaultType && getResolvedDefaultWorker(defaultType);
+                if (!popTypeResolution()) {
+                    return typeParameter.resolvedDefault = circularConstraintOrDefaultType;
```
I removed this following our discussion regarding circular references here. Instead, now when a type parameter default points to an earlier type parameter, it gets the type argument supplied for that parameter or the default selected for that type parameter. If the type parameter default points to a later type parameter it becomes `emptyObjectType`

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -18323,16 +18478,110 @@ namespace ts {
-                        error(node.name, Diagnostics.All_declarations_of_0_must_have_identical_type_parameters, node.name.text);
+
+            // Resolve the type parameters and minimum type argument count for all declarations
+            resolveTypeParametersOfClassOrInterface(symbol);
```
I think we need to sit down and have you explain this logic to me. I'm not getting all the different `typeParameters` on symbols vs. types and why a different resolution and caching scheme is needed on top of what is already there,

rbuckton replied:
This should be much simpler and relatively straightforward now based on our offline discussion.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -865,6 +865,11 @@ namespace ts {
                             }
+                            // Only perform additional check if error reporting was requested
+                            if (nameNotFoundMessage && !isTypeParameterSymbolDeclaredInContainer(result, location)) {
+                                error(errorLocation, Diagnostics.Type_parameter_0_cannot_be_referenced_outside_of_the_declaration_that_defines_it, symbolToString(result));
```
In the design meeting we discussed this being an error, however I can see the benefits of just silently skipping the type parameter as well.

rbuckton commented:
I have looked into including #13609 in this change, but it results in significant change in behavior that may be too much to take on if we want to take this for 2.2.

@ahejlsberg Do you have any other feedback for this change?

MRamos62786 commented:
@rbuckton You mentioned taking this for 2.2 in an earlier comment but I see that this issue still has the 2.3 milestone attached. Will this be in 2.2?

mhegazy commented:
> @rbuckton You mentioned taking this for 2.2 in an earlier comment but I see that this issue still has the 2.3 milestone attached. Will this be in 2.2?

No. this is scheduled for [TS 2.3](https://github.com/Microsoft/TypeScript/wiki/Roadmap#23-may-2017)

jeffijoe commented:
Can't wait for this, writing a library in TS and this will save the consumers from a lot of confusion. Great work!
